sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -107,191 +108,166 @@ mod checked {
-                    invariant_violation!(
-                        "Transaction input checker should check that there is enough gas"
-                    );
-                };
-                coin.balance = Balance::new(new_balance);
-                gas
-            } else {
-                InputValue {
-                    object_metadata: None,
-                    inner: ResultValue {
-                        last_usage_kind: None,
-                        value: None,
-                    },
-                }
-            };
-            let native_extensions = new_native_extensions(
-                state_view.as_child_resolver(),
-                input_object_map,
-                !gas_charger.is_unmetered(),
-                protocol_config,
-                metrics.clone(),
-                tx_context.epoch(),
-            );
-
-            // Set the profiler if in CLI
-            #[skip_checked_arithmetic]
-            move_vm_profiler::tracing_feature_enabled! {
-                use move_vm_profiler::GasProfiler;
-                use move_vm_types::gas::GasMeter;
-
-                let tx_digest = tx_context.digest();
-                let remaining_gas: u64 =
-                    move_vm_types::gas::GasMeter::remaining_gas(gas_charger.move_gas_status())
-                        .into();
-                gas_charger
-                    .move_gas_status_mut()
-                    .set_profiler(GasProfiler::init(
-                        &vm.config().profiler_config,
-                        format!("{}", tx_digest),
-                        remaining_gas,
-                    ));
-            }
+    /// A `LinkedContext` is an execution context with a specific linkage that has been determined.
+    /// This is usually for a specific command, but the linkage can be derived, e.g., for the
+    /// entire PTB.
+    pub struct LinkedContext<'ctx, 'vm, 'state, 'a> {
+        /// The execution context
+        pub ctx: &'ctx mut ExecutionContext<'vm, 'state, 'a>,
+        pub vm_instance: MoveVM<'state>,
+        /// The specified linkage for this linked context.
+        /// This is a mapping of runtime_id -> storage_id
+        pub linkage: ResolvedLinkage,
+        /// A "reverse" linkage of storage_id -> runtime_id
+        pub reverse_linkage: BTreeMap<ObjectID, ObjectID>,
+    }
 
+    impl<'ctx, 'vm, 'state, 'a> LinkedContext<'ctx, 'vm, 'state, 'a> {
+        pub fn new(
+            ctx: &'ctx mut ExecutionContext<'vm, 'state, 'a>,
+            linkage: ResolvedLinkage,
+        ) -> VMResult<Self> {
+            let vm_instance = ctx.vm.make_vm_with_native_extensions(
+                SuiDataStore::new(&ctx.state_view.as_sui_resolver(), &ctx.new_packages),
+                into_linkage_context(linkage.clone()),
+                ctx.native_extensions.clone(),
+            )?;
+            let reverse_linkage = linkage
+                .iter()
+                .map(|(k, v)| (*v, *k))
```

Comment:
Is this guaranteed to be bidirectional?

Reply:
Yes, we are guaranteed that every version of a package corresponds to exactly one original package id (i.e., package UUID).

Basically, if there were a conflict here it would mean that package `A@v1` is somehow both a `A@v1` and a `B@v_something` package which is impossible.